compiler/rustc_ast/src/ast.rs

```diff
@@ -323,6 +325,63 @@ impl GenericBound {
 
 pub type GenericBounds = Vec<GenericBound>;
 
+struct LifetimeCollectVisitor<'ast> {
+    current_binders: Vec<NodeId>,
+    binders_to_ignore: FxHashMap<NodeId, Vec<NodeId>>,
+    collected_lifetimes: Vec<&'ast Lifetime>,
+}
+
+impl<'ast> Visitor<'ast> for LifetimeCollectVisitor<'ast> {
+    fn visit_lifetime(&mut self, lifetime: &'ast Lifetime, _: LifetimeCtxt) {
+        if !self.collected_lifetimes.contains(&lifetime) {
```

Comment:
I expect this code to say

```rust
if !self.current_binders.contains(lifetime) {
    if !self.collected_lifetimes.contains(lifetime) {
        self.collected_lifetimes.push(lifetime);
    }
}
```

i.e., just don't collect the lifetime if it is bound.

Reply:
You need the `LifetimeRes` to tell where the lifetime was introduced.
```
let binder = match res {
    LifetimeRes::Param { binder, .. } | LifetimeRes::Fresh { binder, .. } => Some(binder),
    _ => None,
};
if binder.map_or(true, |b| !self.current_binders.contains(b)) {
```

Reply:
Fixed https://github.com/rust-lang/rust/pull/99867/commits/7915f15c1c007e23e7cd0b152914fe0ddb808915